src/api/mod.rs

```diff
@@ -174,6 +184,112 @@ impl Api {
             }
         });
     }
+
+    fn start_eth_background_task(
+        &self,
+        head_tx: watch::Sender<Option<(JsonValue, u64)>>,
+        finalized_head_tx: watch::Sender<Option<(JsonValue, u64)>>,
+    ) {
+        let client = self.client.clone();
+        let stale_timeout = self.stale_timeout;
+
+        tokio::spawn(async move {
+            let mut interval = tokio::time::interval(stale_timeout);
+
+            loop {
+                let client = client.clone();
+
+                let run = async {
+                    interval.reset();
+
+                    // query current head
+                    let head = client
+                        .request("eth_getBlockByNumber", vec!["latest".into(), true.into()])
+                        .await?;
+                    let number = get_number(&head)?;
+                    let hash: JsonValue = head["hash"].clone();
+
+                    tracing::debug!("New head: {number} {hash}");
+                    head_tx.send_replace(Some((hash.clone(), number)));
+
+                    let mut sub = client
+                        .subscribe(
+                            "eth_subscribe",
+                            ["newHeads".into()].into(),
+                            "eth_unsubscribe",
+                        )
+                        .await?;
+
+                    loop {
+                        tokio::select! {
+                            val = sub.next() => {
+                                if let Some(Ok(val)) = val {
+                                    interval.reset();
+
+                                    let number = get_number(&val)?;
+
+                                    let hash: JsonValue = val["hash"].clone();
+
+                                    tracing::debug!("New head: {number} {hash}");
+                                    head_tx.send_replace(Some((hash.clone(), number)));
+                                } else {
+                                    break;
+                                }
+                            }
+                            _ = interval.tick() => {
+                                tracing::warn!("No new blocks for {stale_timeout:?} seconds, rotating endpoint");
+                                client.rotate_endpoint().await.expect("Failed to rotate endpoint");
+                                break;
+                            }
+                        }
+                    }
+
+                    Ok::<(), anyhow::Error>(())
+                };
+
+                if let Err(e) = run.await {
+                    tracing::error!("Error in background task: {e}");
+                }
+                tokio::time::sleep(Duration::from_secs(1)).await;
+            }
+        });
+
+        if !self.eth_subscribe_finalized {
```

Comment:
I don't think we need to make this configurable.
Just subscribe to it and handle the error.

Reply:
yeah that works as well